CatVariants can mis-sort input VCFs without -assumeSorted

johnwallace123 opened the pull request:
When using CatVariants, VCF files were being sorted solely on the base
pair position of the first record, ignoring the chromosome.  This can
become problematic when merging files from different chromosomes,
espeically if you have multiple VCFs per chromosome.

As an example, assume the following 3 lines are all in separate files:
1       10
1       100
2       20

The merged VCF from CatVariants (without -assumeSorted) would read:
1       10
2       20
1       100

This has the potential to break tools that expect chromosomes to be
contiguous within a VCF file.

This commit changes the comparator from one of Pair\<Integer, File\> to
one of Pair\<VariantContext, File\>.  We construct a
VariantContextComparator from the provided reference, which will sort
the first record by chromosome and position properly.  Additionally, if
-assumeSorted is given, we simply use a null VariantContext as the first
record, which will all be equal (as all will be null)

vdauwera commented:
@johnwallace123 Sorry it's taken us a while to get to this! We're processing it now.

ronlevine commented on public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java:
```diff
@@ -318,15 +320,22 @@ public static void main(String[] args){
+            return comp.compare(startPositionP1, startPositionP2);
+            //if (startPositionP1  == startPositionP2)
+            //    return 0;
+            //return startPositionP1 < startPositionP2 ? -1 : 1 ;
```
Remove commented out lines

ronlevine replied:
Add a test to CatVariantsIntegrationTes for checking the sorting logic/comparator.

ronlevine commented on public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java:
```diff
@@ -257,9 +259,9 @@ protected int execute() throws Exception {
                 }
                 VariantContext vc = it.next();
-                int firstPosition = vc.getStart();
+                //int firstPosition = vc.getStart();
```
Remove commented out line.

ronlevine commented on public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java:
```diff
@@ -318,15 +320,22 @@ public static void main(String[] args){
+    	
+    	VariantContextComparator comp;
+    	
+    	public PositionComparator(SAMSequenceDictionary dict){
```
make argument "final".

ronlevine commented on public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java:
```diff
@@ -318,15 +320,22 @@ public static void main(String[] args){
-            if (startPositionP1  == startPositionP2)
-                return 0;
-            return startPositionP1 < startPositionP2 ? -1 : 1 ;
+        public int compare(Pair<VariantContext,File> p1, Pair<VariantContext,File> p2) {
```
Make arguments "final".

johnwallace123 commented:
@ronlevine I believe I've added all the changes that you requested.  If you approve, I'll rebase into a single commit, otherwise I'll keep working :).

On a related note, I wasn't able to actually get any tests to run using "mvn verify" or "mvn integration-test"; it would always say "Tests are skipped".  I verified by adding an Assert.fail() into the test that I added.  Is there something that I'm doing wrong?

Thanks!

vdauwera commented:
@johnwallace123 Not your fault, the test framework isn't available in the public source. We know that sucks -- will be done differently in GATK4. Apologies if you spent time trying to get the tests to work.

ronlevine commented:
@johnwallace123 Don't worry about the tests. I already cherry-picked your changes, put them on a branch in a Broad private repo and created an integration test for your code. The command to run the integration test is:
./ant-bridge.sh integrationtest -Dsingle=CatVariantsIntegrationTest
But it will not run successfully for you because it needs to access Broad internal data

johnwallace123 commented:
@vdauwera @ronlevine:  Thanks for confirming that I wasn't going crazy!  If you need anything else from me, please let me know.
